File: leetcode/leetcode75/array_string/238_product_of_array_except_self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        n = len(nums)
        res = [1]*n
        for i in range(1, n):
            res[i] = res[i-1]*nums[i-1]
        right = 1
        for i in reversed(range(0, n)):
            res[i] *= right
            right *= nums[i]
            print(f'i: {i}\tres[i]:{res[i]}')
        return res

if __name__ == "__main__":
    s = Solution()
    nums = [1,2,3,4]
    print(s.productExceptSelf(nums))

File: NeetCode RoadMap/arrays&hashing/238_Product_of_Array_Except_Self.cpp
#include<vector>
#include<iostream>

using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> sol(nums.size(), 1);
        int right = nums[nums.size()-1];
        for(int i=1; i<nums.size(); i++) { //i=1 since there's no number before first
            sol[i] = sol[i-1] * nums[i-1]; //first multiply with all the numbers on left.
        }

        for(int j = nums.size()-2; j>=0; j--) {//keep a track of the right side product until current, sol[i] is the product of all elements towards left of i.
            sol[j] = sol[j]*right;
            right *= nums[j];
        }

        return sol;
    }
};

int main() {
    Solution sol;
    vector<int> nums = {1, 2, 3, 4};
    vector<int> res = sol.productExceptSelf(nums);
    for(auto it: res) {
        cout << it << " ";
    }
    return 0;
}